### PR TITLE
The storage engine type awareness in Kvrocks

### PR DESCRIPTION
**Pull request by MaheshMadushan**
https://github.com/apache/kvrocks/issues/1966

Storage Engine Type Enum
`enum class StorageEngineType : uint16_t {
  RocksDB,
  Speedb,
};`

Storage type stored in variable STORAGE_ENGINE_TYPE in `src/storage.h`. it can be `StorageEngineType::RocksDB` or `StorageEngineType::Speedb`

### CODE REVIEW

**Review by git-hulk — DISMISSED** [review]
LGTM, thank you!

File: CMakeLists.txt [hunk 1]
```diff
@@ -216,6 +216,11 @@ elseif((CMAKE_CXX_COMPILER_ID STREQUAL "Clang") OR (CMAKE_CXX_COMPILER_ID STREQU
 endif()
 target_link_libraries(kvrocks_objs PUBLIC -fno-omit-frame-pointer)
 target_link_libraries(kvrocks_objs PUBLIC ${EXTERNAL_LIBS})
+if (ENABLE_SPEEDB)
+    target_compile_definitions(kvrocks_objs PUBLIC SPEEDB_STORAGE_ENGINE)
```
Comment (PragmaTwice) [inline CMakeLists.txt]: Could we just use one macro definition with different value?

**Review by PragmaTwice — CHANGES_REQUESTED** [review]
BTW, I think we can have an enum type in our code.

File: CMakeLists.txt (hunk 1 above)
Comment (MaheshMadushan) [reply CMakeLists.txt]: Thank you for the feedback. I'm thinking of having numbered values for each storage engine type. Would that be ideal?
or we could have string names like `"ROCKSDB"` and `"SPEEDB"`.

- First case:

`STORAGE_ENGINE=0 # for rocksdb`
`STORAGE_ENGINE=1 # for speedb`

- Second case:

`STORAGE_ENGINE="ROCKSDB"`
`STORAGE_ENGINE="SPEEDB"`

Comment (mapleFU) [reply CMakeLists.txt]: I vote for the second one

**Comment by MaheshMadushan** [comment]
> BTW, I think we can have an enum type in our code.

Under the namespace `engine` in `src/storage.h` following enum can be created.
`enum class Type {
  RocksDB,
  Speedb,
};
`

Then we can have macros for storage engine as following,
`STORAGE_ENGINE=0 # for rocksdb`
`STORAGE_ENGINE=1 # for speedb`

Then we can determine the engine type as following,
`if (STORAGE_ENGINE == static_cast<int>(engine::Type::RocksDB)) {} ...`

Would that be ideal?

**Comment by PragmaTwice** [comment]
> > BTW, I think we can have an enum type in our code.
> 
> Under the namespace `engine` in `src/storage.h` following enum can be created.
> `enum class Type {
>   RocksDB,
>   Speedb,
> };
> `
> 
> Then we can have macros for storage engine as following,
> `STORAGE_ENGINE=0 # for rocksdb`
> `STORAGE_ENGINE=1 # for speedb`
> 
> Then we can determine the engine type as following,
> `if (STORAGE_ENGINE == static_cast<int>(engine::Type::RocksDB)) {} ...`
> 
> Would that be ideal?

Firstly, the enum type can definitely have a better name than `Type`.

Besides, we can have a static-storage-duration constant to wrap the macro, so that we don't need to do any type cast.

File: CMakeLists.txt [hunk 2]
```diff
@@ -31,6 +31,8 @@ option(ENABLE_IPO "enable interprocedural optimization" ON)
 option(ENABLE_UNWIND "enable libunwind in glog" ON)
 option(ENABLE_SPEEDB "enable speedb instead of rocksdb" OFF)
 set(PORTABLE 0 CACHE STRING "build a portable binary (disable arch-specific optimizations)")
+set(ROCKSDB_STORAGE_ENGINE 0 STRING "ROCKSDB")
+set(SPEEDB_STORAGE_ENGINE 1 STRING "SPEEDB")
```
Comment (PragmaTwice) [inline CMakeLists.txt:35]:
```suggestion
```

Comment (PragmaTwice) [inline CMakeLists.txt]:
```suggestion
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -216,6 +218,11 @@ elseif((CMAKE_CXX_COMPILER_ID STREQUAL "Clang") OR (CMAKE_CXX_COMPILER_ID STREQU
 endif()
 target_link_libraries(kvrocks_objs PUBLIC -fno-omit-frame-pointer)
 target_link_libraries(kvrocks_objs PUBLIC ${EXTERNAL_LIBS})
+if (ENABLE_SPEEDB)
+    target_compile_definitions(kvrocks_objs PUBLIC STORAGE_ENGINE=${SPEEDB_STORAGE_ENGINE})
+else()
+    target_compile_definitions(kvrocks_objs PUBLIC STORAGE_ENGINE=${ROCKSDB_STORAGE_ENGINE})
+endif()
```
Comment (PragmaTwice) [inline CMakeLists.txt]:
```suggestion
if (ENABLE_SPEEDB)
    target_compile_definitions(kvrocks_objs PUBLIC KVROCKS_STORAGE_ENGINE=Speedb)
else()
    target_compile_definitions(kvrocks_objs PUBLIC KVROCKS_STORAGE_ENGINE=RocksDB)
endif()
```

File: src/storage/storage.h [hunk 4]
```diff
@@ -40,6 +40,13 @@
 #include "observer_or_unique.h"
 #include "status.h"
 
+enum class StorageEngineType : uint16_t {
+  RocksDB,
+  Speedb,
+};
+
+static constexpr StorageEngineType STORAGE_ENGINE_TYPE = StorageEngineType { STORAGE_ENGINE };
```
Comment (PragmaTwice) [inline src/storage/storage.h]:
```suggestion
inline constexpr StorageEngineType STORAGE_ENGINE_TYPE = StorageEngineType::KVROCKS_STORAGE_ENGINE;
```

**Review by PragmaTwice — APPROVED** [review]

**Comment by PragmaTwice** [comment]
Thank you for your contribution!

**Pull request merged**